File: Dopamine.CollectionModule/ViewModels/CollectionTracksColumnsViewModel.vb
Imports Dopamine.Core.Settings
Imports Microsoft.Practices.Prism.Commands
Imports Microsoft.Practices.Prism.Mvvm

Namespace ViewModels
    Public Class CollectionTracksColumnsViewModel
        Inherits BindableBase

#Region "Private"
        Private mCheckBoxRatingVisible As Boolean

        Private mCheckBoxRatingChecked As Boolean
        Private mCheckBoxArtistChecked As Boolean
        Private mCheckBoxAlbumChecked As Boolean
        Private mCheckBoxGenreChecked As Boolean
        Private mCheckBoxLengthChecked As Boolean
        Private mCheckBoxAlbumArtistChecked As Boolean
        Private mCheckBoxTrackNumberChecked As Boolean
        Private mCheckBoxYearChecked As Boolean
#End Region

#Region "Properties"
        Public Property CheckBoxRatingVisible() As Boolean
            Get
                Return mCheckBoxRatingVisible
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mCheckBoxRatingVisible, value)
            End Set
        End Property

        Public Property CheckBoxRatingChecked() As Boolean
            Get
                Return mCheckBoxRatingChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mCheckBoxRatingChecked, value)
            End Set
        End Property

        Public Property CheckBoxArtistChecked() As Boolean
            Get
                Return mCheckBoxArtistChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mCheckBoxArtistChecked, value)
            End Set
        End Property

        Public Property CheckBoxAlbumChecked() As Boolean
            Get
                Return mCheckBoxAlbumChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mCheckBoxAlbumChecked, value)
            End Set
        End Property

        Public Property CheckBoxGenreChecked() As Boolean
            Get
                Return mCheckBoxGenreChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mCheckBoxGenreChecked, value)
            End Set
        End Property

        Public Property CheckBoxLengthChecked() As Boolean
            Get
                Return mCheckBoxLengthChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mCheckBoxLengthChecked, value)
            End Set
        End Property

        Public Property CheckBoxAlbumArtistChecked() As Boolean
            Get
                Return mCheckBoxAlbumArtistChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mCheckBoxAlbumArtistChecked, value)
            End Set
        End Property

        Public Property CheckBoxTrackNumberChecked() As Boolean
            Get
                Return mCheckBoxTrackNumberChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mCheckBoxTrackNumberChecked, value)
            End Set
        End Property

        Public Property CheckBoxYearChecked() As Boolean
            Get
                Return mCheckBoxYearChecked
            End Get
            Set(ByVal value As Boolean)
                SetProperty(Of Boolean)(Me.mCheckBoxYearChecked, value)
            End Set
        End Property
#End Region

#Region "Commands"
        Public Property OkCommand As DelegateCommand
        Public Property CancelCommand As DelegateCommand
#End Region

#Region "Construction"
        Public Sub New()

            If XmlSettingsClient.Instance.Get(Of Boolean)("Behaviour", "EnableRating") Then
                Me.CheckBoxRatingVisible = True
            Else
                Me.CheckBoxRatingVisible = False
            End If

            Me.GetVisibleColumns()
        End Sub
#End Region

#Region "Private"
        Private Sub GetVisibleColumns()
            Utils.GetVisibleSongsColumns(Me.CheckBoxRatingChecked,
                                         Me.CheckBoxArtistChecked,
                                         Me.CheckBoxAlbumChecked,
                                         Me.CheckBoxGenreChecked,
                                         Me.CheckBoxLengthChecked,
                                         Me.CheckBoxAlbumArtistChecked,
                                         Me.CheckBoxTrackNumberChecked,
                                         Me.CheckBoxYearChecked)
        End Sub
#End Region

#Region "Public"
        Public Async Function SetVisibleColumns() As Task(Of Boolean)

            Await Task.Run(Sub()
                               Utils.SetVisibleSongsColumns(Me.CheckBoxRatingChecked,
                                                            Me.CheckBoxArtistChecked,
                                                            Me.CheckBoxAlbumChecked,
                                                            Me.CheckBoxGenreChecked,
                                                            Me.CheckBoxLengthChecked,
                                                            Me.CheckBoxAlbumArtistChecked,
                                                            Me.CheckBoxTrackNumberChecked,
                                                            Me.CheckBoxYearChecked)
                           End Sub)

            Return True
        End Function
#End Region
    End Class
End Namespace

File: Dopamine.CollectionModule/ViewModels/CollectionTracksColumnsViewModel.cs
using Digimezzo.Utilities.Settings;
using Prism.Commands;
using Prism.Mvvm;
using System.Threading.Tasks;

namespace Dopamine.CollectionModule.ViewModels
{
    public class CollectionTracksColumnsViewModel : BindableBase
    {
        #region Private
        private bool checkBoxRatingVisible;
        private bool checkBoxLoveVisible;
        private bool checkBoxRatingChecked;
        private bool checkBoxLoveChecked;
        private bool checkBoxLyricsChecked;
        private bool checkBoxArtistChecked;
        private bool checkBoxAlbumChecked;
        private bool checkBoxGenreChecked;
        private bool checkBoxLengthChecked;
        private bool checkBoxPlayCountChecked;
        private bool checkBoxSkipCountChecked;
        private bool checkBoxDateLastPlayedChecked;
        private bool checkBoxAlbumArtistChecked;
        private bool checkBoxTrackNumberChecked;
        private bool checkBoxYearChecked;
        private bool checkBoxBitrateChecked;
        #endregion

        #region Properties
        public bool CheckBoxRatingVisible
        {
            get { return checkBoxRatingVisible; }
            set { SetProperty<bool>(ref this.checkBoxRatingVisible, value); }
        }

        public bool CheckBoxLoveVisible
        {
            get { return checkBoxLoveVisible; }
            set { SetProperty<bool>(ref this.checkBoxLoveVisible, value); }
        }

        public bool CheckBoxRatingChecked
        {
            get { return checkBoxRatingChecked; }
            set { SetProperty<bool>(ref this.checkBoxRatingChecked, value); }
        }

        public bool CheckBoxLoveChecked
        {
            get { return checkBoxLoveChecked; }
            set { SetProperty<bool>(ref this.checkBoxLoveChecked, value); }
        }

        public bool CheckBoxLyricsChecked
        {
            get { return checkBoxLyricsChecked; }
            set { SetProperty<bool>(ref this.checkBoxLyricsChecked, value); }
        }

        public bool CheckBoxArtistChecked
        {
            get { return checkBoxArtistChecked; }
            set { SetProperty<bool>(ref this.checkBoxArtistChecked, value); }
        }

        public bool CheckBoxAlbumChecked
        {
            get { return checkBoxAlbumChecked; }
            set { SetProperty<bool>(ref this.checkBoxAlbumChecked, value); }
        }

        public bool CheckBoxGenreChecked
        {
            get { return checkBoxGenreChecked; }
            set { SetProperty<bool>(ref this.checkBoxGenreChecked, value); }
        }

        public bool CheckBoxLengthChecked
        {
            get { return checkBoxLengthChecked; }
            set { SetProperty<bool>(ref this.checkBoxLengthChecked, value); }
        }

        public bool CheckBoxPlayCountChecked
        {
            get { return checkBoxPlayCountChecked; }
            set { SetProperty<bool>(ref this.checkBoxPlayCountChecked, value); }
        }

        public bool CheckBoxSkipCountChecked
        {
            get { return checkBoxSkipCountChecked; }
            set { SetProperty<bool>(ref this.checkBoxSkipCountChecked, value); }
        }

        public bool CheckBoxDateLastPlayedChecked
        {
            get { return checkBoxDateLastPlayedChecked; }
            set { SetProperty<bool>(ref this.checkBoxDateLastPlayedChecked, value); }
        }

        public bool CheckBoxAlbumArtistChecked
        {
            get { return checkBoxAlbumArtistChecked; }
            set { SetProperty<bool>(ref this.checkBoxAlbumArtistChecked, value); }
        }

        public bool CheckBoxTrackNumberChecked
        {
            get { return checkBoxTrackNumberChecked; }
            set { SetProperty<bool>(ref this.checkBoxTrackNumberChecked, value); }
        }

        public bool CheckBoxYearChecked
        {
            get { return checkBoxYearChecked; }
            set { SetProperty<bool>(ref this.checkBoxYearChecked, value); }
        }

        public bool CheckBoxBitrateChecked
        {
            get { return checkBoxBitrateChecked; }
            set { SetProperty<bool>(ref this.checkBoxBitrateChecked, value); }
        }
        #endregion

        #region Commands
        public DelegateCommand OkCommand { get; set; }
        public DelegateCommand CancelCommand { get; set; }
        #endregion

        #region Construction
        public CollectionTracksColumnsViewModel()
        {
            this.CheckBoxRatingVisible = SettingsClient.Get<bool>("Behaviour", "EnableRating");
            this.CheckBoxLoveVisible = SettingsClient.Get<bool>("Behaviour", "EnableLove");

            this.GetVisibleColumns();
        }
        #endregion

        #region Private
        private void GetVisibleColumns()
        {
            Utils.GetVisibleSongsColumns(
                ref this.checkBoxRatingChecked,
                ref this.checkBoxLoveChecked,
                ref this.checkBoxLyricsChecked,
                ref this.checkBoxArtistChecked,
                ref this.checkBoxAlbumChecked,
                ref this.checkBoxGenreChecked,
                ref this.checkBoxLengthChecked,
                ref this.checkBoxPlayCountChecked,
                ref this.checkBoxSkipCountChecked,
                ref this.checkBoxDateLastPlayedChecked,
                ref this.checkBoxAlbumArtistChecked,
                ref this.checkBoxTrackNumberChecked,
                ref this.checkBoxYearChecked,
                ref this.checkBoxBitrateChecked);

            OnPropertyChanged(() => this.CheckBoxRatingChecked);
            OnPropertyChanged(() => this.CheckBoxLoveChecked);
            OnPropertyChanged(() => this.CheckBoxLyricsChecked);
            OnPropertyChanged(() => this.CheckBoxArtistChecked);
            OnPropertyChanged(() => this.CheckBoxAlbumChecked);
            OnPropertyChanged(() => this.CheckBoxGenreChecked);
            OnPropertyChanged(() => this.CheckBoxLengthChecked);
            OnPropertyChanged(() => this.CheckBoxPlayCountChecked);
            OnPropertyChanged(() => this.CheckBoxSkipCountChecked);
            OnPropertyChanged(() => this.CheckBoxDateLastPlayedChecked);
            OnPropertyChanged(() => this.CheckBoxAlbumArtistChecked);
            OnPropertyChanged(() => this.CheckBoxTrackNumberChecked);
            OnPropertyChanged(() => this.CheckBoxYearChecked);
            OnPropertyChanged(() => this.CheckBoxBitrateChecked);
        }
        #endregion

        #region Public
        public async Task<bool> SetVisibleColumns()
        {
            await Task.Run(() =>
            {
                Utils.SetVisibleSongsColumns(
                    this.CheckBoxRatingChecked,
                    this.CheckBoxLoveChecked,
                    this.CheckBoxLyricsChecked,
                    this.CheckBoxArtistChecked,
                    this.CheckBoxAlbumChecked,
                    this.CheckBoxGenreChecked,
                    this.CheckBoxLengthChecked,
                    this.CheckBoxPlayCountChecked,
                    this.CheckBoxSkipCountChecked,
                    this.CheckBoxDateLastPlayedChecked,
                    this.CheckBoxAlbumArtistChecked,
                    this.CheckBoxTrackNumberChecked,
                    this.CheckBoxYearChecked,
                    this.checkBoxBitrateChecked);
            });

            return true;
        }
        #endregion
    }
}
